File: src/main/java/com/plugin/frege/stubs/index/FregeClassNameIndex.java
package com.plugin.frege.stubs.index;

import com.intellij.psi.stubs.StringStubIndexExtension;
import com.intellij.psi.stubs.StubIndexKey;
import com.plugin.frege.psi.FregePsiClass;
import org.jetbrains.annotations.NotNull;

public class FregeClassNameIndex extends StringStubIndexExtension<FregePsiClass> {
    private static final StubIndexKey<String, FregePsiClass> KEY =
            StubIndexKey.createIndexKey("com.plugin.frege.stubs.index.FregeClassNameIndex");
    private static final FregeClassNameIndex INSTANCE = new FregeClassNameIndex();

    private FregeClassNameIndex() { }

    public static FregeClassNameIndex getInstance() {
        return INSTANCE;
    }

    @Override
    public @NotNull StubIndexKey<String, FregePsiClass> getKey() {
        return KEY;
    }
}

File: src/main/kotlin/com/plugin/frege/stubs/index/FregeClassNameIndex.kt
package com.plugin.frege.stubs.index

import com.intellij.psi.stubs.StubIndexKey
import com.plugin.frege.psi.FregePsiClass

object FregeClassNameIndex : FregeNameIndex<FregePsiClass>() {
    private val KEY = StubIndexKey.createIndexKey<Int, FregePsiClass>(
        "com.plugin.frege.stubs.index.FregeClassNameIndex"
    )

    override fun getKey(): StubIndexKey<Int, FregePsiClass> = KEY

    override fun nameMatched(element: FregePsiClass, name: String): Boolean {
        return element.qualifiedName == name
    }
}
